### PR TITLE
ARROW-17106: [Python] Move init code to core and expose only API

### PR DESCRIPTION
**Pull request by milesgranger**
Will close [ARROW-17106](https://issues.apache.org/jira/browse/ARROW-17106)

Moves all current code in `pyarrow/parquet/__init__.py` to `pyarrow/parquet/core.py` and re-exports the public API from core.py inside of the init file.

### CODE REVIEW

**Comment by milesgranger** [comment]
Opened [ARROW-17320](https://issues.apache.org/jira/browse/ARROW-17320) to refine which symbols should be designated as private/public from `pyarrow.parquet`. For now, to keep same functionality, re-export all from `core.py`

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

**Comment by pitrou** [comment]
@github-actions crossbow submit -g python

File: python/pyarrow/parquet/core.py [hunk 1]
```diff
@@ -0,0 +1,3467 @@
...
+    return ParquetFile(
+        where, memory_map=memory_map,
+        decryption_properties=decryption_properties).schema.to_arrow_schema()
+
+
+# re-export everything
+# std `from . import *` ignores symbols with leading `_`
+__all__ = [
+    m for m, _ in inspect.getmembers(sys.modules[__name__])
+]
```
Comment (pitrou) [inline python/pyarrow/parquet/core.py:3467]: It seems this is the same as the following, but much slower?
```suggestion
__all__ = list(__dict__)
```

**Review by pitrou — COMMENTED** [review]
Thanks for doing this. Hopefully we can soon define a curated set of exported APIs.

File: python/pyarrow/parquet/core.py (hunk 1 above)
Comment (milesgranger) [reply python/pyarrow/parquet/core.py]: Maybe I'm missing somehting, but I get 
```
-> 3463 __all__ = list(__dict__)

NameError: name '__dict__' is not defined
```

Changing it to `list(sys.modules[__name__].__dict__)` and a smidgen faster. Thanks!

```python
In [10]: N = 1_000

In [11]: # __dict__

In [12]: timeit.timeit('_ = importlib.reload(core)', number=N, globals=globals()) / N
Out[12]: 0.0005946487420005723

In [13]: # inspect.getmembers

In [14]: timeit.timeit('_ = importlib.reload(core)', number=N, globals=globals()) / N
Out[14]: 0.000692891882998083

In [15]: 
```

Comment (pitrou) [reply python/pyarrow/parquet/core.py]: By the way, since you're using IPython, you probably know it but the `%timeit` macro is useful :-)

Comment (milesgranger) [reply python/pyarrow/parquet/core.py]: Recently switched to it actually, thanks for the tip! 👍

**Comment by pitrou** [comment]
Hmm, the crash on the Sphinx & Numpydoc CI job is unfortunate but certainly not related I think.

**Pull request merged**